hazelcast/src/main/java/com/hazelcast/util/collection/OAHashSet.java

```diff
@@ -103,8 +101,17 @@ public OAHashSet(int initialCapacity) {
      * @param loadFactor      the load factor of the set to be created
      */
     public OAHashSet(int initialCapacity, float loadFactor) {
-        checkNotNegative(initialCapacity, "Illegal initial capacity: " + initialCapacity);
-        checkFalse(loadFactor <= 0 || loadFactor >= 1 || Float.isNaN(loadFactor), "Illegal load factor: " + loadFactor);
+        // the parameter checks below are intentionally not done via Preconditions
+        // the error messages provided to the preconditions are created unconditionally
+        // which creates plenty StringBuilders and for building the error message
+        // if many instances are created in a loop this increases the GC pressure significantly
+        if (initialCapacity < 0) {
+            throw new IllegalArgumentException("Illegal initial capacity: " + initialCapacity);
+        }
+
+        if (loadFactor <= 0 || loadFactor >= 1 || Float.isNaN(loadFactor)) {
+            throw new IllegalArgumentException("Illegal load factor: " + loadFactor);
+        }
```

Comment:
@mmedenjak @tkountis :point_up: 
This change brought the speedup.

Reply:
Damn Java 6 :) 
However, we could slightly improve the preconditions too, with formatted strings and varargs.

Reply:
Varargs are evil! Should not be used because you get an array and boxing.. 

I'll put it on my list to check all Preconditions methods. They should have no litter.

Reply:
The issue here was the unconditional call-site error message construction, not the `Preconditions` method internals. Same issue as with the complex log messages if not guarded by `isLevelEnabled()` checks. Creating this error message: `"Illegal load factor: " + loadFactor` * `partition count` * `merkle tree leaf count` put the heap and GC under heavy fire.

Reply:
What I wanted to say is introduce the same principle as with logging frameworks. eg. have methods with 1,2,3,4... arguments, and then varargs as the last resort. This way we can support formatted strings. But Peter, might have better ideas.